### ISSUE TITLE
Testing for components that make use of integrations: gitto, tasks, nomad

### ISSUE THREAD

**Issue by nikochiko** [pipalacademy/capstone#57]
These include:
- Creating a new project
- Updating an existing project
- Starting a user project
- Updating a user project
  - Deploying a project
  - Running checks for a project

For each test, several things need to be working at the same time.

For example in updating a user project,
it requires the task worker (redis and rq worker), nomad, docker, and gitto to be working.

These tests must happen in the CI.

**Comment by nikochiko** [comment]
An important thing to note is that the tests must use a different namespace inside each of the services.

e.g. different namespace in nomad, different database in postgres, a different temporary deployment of gitto and docker-registry, a separate queue on redis.

Due to conflicting ports in nomad, the tests must use ports prepended by 1 (e.g. 18080 instead of 8080, 17878 instead of 7878). this can be configured with nomad variables. The worker can start listening on a separate queue for tests. The worker must be started by the test script to run in the background. its logs can be pushed to a task-test.log file or something similarly named. it will be overwritten on each run because it is not useful unless the tests fail.
postgres and redis are assumed to be running.
this setup can be replicated on github ci as well.

## Project creation

This is what happens in the real app:
1. Server-admin runs the CLI command `capstone-server projects new`, grabs `git_url` from the output.
2. Project author clones the repository
4. Server admin publishes the project with `capstone-server projects publish NAME`
5. Project changes must show on the project website
6. When learners start the project and clone it, they must see the content uploaded by author in repo/ directory

These are the test cases:
1. project repo has initial content - README.md, capstone.yml (with correct info), requirements.txt, checks.py, and repo directory
2. project is unpublished by default
3. changes to title/description/tags/short-description work
4. invalid capstone.yml files are rejected (missing fields)
5. project is visible on website after publishing, and learner can start project (involves start-user-project)
6. learner's git repo contains code pushed by author (involves update-user-project also)

## Project updates

1. Project author makes changes, git pushes
2. Cases such as deleting tasks, checks, should be handled. Some of these might involve deleting user's progress. This must be verified that it works correctly (maybe in separate unit tests).
3. Changes must be visible on the webpage
4. When learners clone project, they must see the new content.

Test cases:
1. Author updates project content: README.md, requirements.txt, checks.py, and repo directory
2. Author updates project capstone.yml: such as title, description, tags, and short description
3. Author submits invalid capstone.yml, should be rejected, or author must be alerted
4. Author deletes tasks or checks from the project, and appropriate user progress is deleted as well
5. Changes made by the author are reflected on the project webpage
6. Learners who clone the project after updates see the updated content in their repository
7. For checks: requirements.txt dependencies must be installed
8. What to do when a project is updated when it has learners?

Check for logs/changelog also?

## Start user project

In capstone

1. Learner visits the project page on the website and clicks "Start Project."
2. The backend creates a new user project, clones the base project repository, and sets up the user's environment.
3. The learner receives a unique URL to access their project

Test cases:
1. Verify that a new user project is created with the correct base project repository and initial content.
2. Test that the unique URL generated for each user project is accessible.
3. Check that the user project is properly linked to the learner's account, and progress is saved and tracked. (git url, repo id must match)
4. Deployment must be created on nomad

## Update user project

In Capstone:
1. Learner makes changes to their project files and commits the changes to their repository.
2. The backend builds new docker image, deploys it with nomad

Test cases:
1. Changes made by learner are correctly saved (commit hash of remote and user must match)
2. Deployment must be updated
3. Checks must be run, and result must be updated
4. What happens when user pushes to a different branch from main?

**Comment by anandology** [comment]
We should not need redis and rq worker for running tests.

The [RQ docs on testing][1] recommends using `FakeRedis` with `is_aync=False`. 

[1]: https://python-rq.org/docs/testing/

I think we can create a fake queue in `tasks.py` if `config.test_mode` is enabled.

**Comment by nikochiko** [comment]
Yes, that will be better.
Thanks for the pointer

**Comment by nikochiko** [comment]
To organize integration tests, I am creating a new directory inside `tests/` called `test_integration/`. Before we start writing integration tests, some more code changes will be needed to allow this.

Setup needed for integration testing:
- [ ] CAPSTONE_TEST config variable
- [ ] FakeRedis, task queue with fake redis, fixture for worker to run `SimpleWorker`
- [ ] Nomad: variable namespace for each job. namespace for services created dynamically.
- [ ] Nomad: variable for ports (for default tasks: nginx, gitto, docker-registry). Set these as env vars in start-jobs.sh. Set env vars for capstone in runtests.sh

Questions:
- How to auto-delete nomad after each test? Easy fix: delete the `test` namespace before start of each test run.

**Comment by nikochiko** [comment]
Roadblock while testing projects.

Click's recommended testing method sets stdout, stderr to some other value (like `IOString`) to capture its value. This doesn't work well when the command starts a subprocess that itself writes to stdout/stderr (subprocess module tries to pass the current stdout/stderr to the subprocess by using the `.fileno()` method, but IOString doesn't have that method).

**Comment by nikochiko** [comment]
Problem: how to run tests where gitto needs to send a webhook request to the app. blocker: gitto would run inside nomad and by default it won't be able to access our app if our app runs on host because it would be on a different network interface.

- Mocking webhook calls ourselves won't let us test whether gitto is actually making those webhook calls or not. There is an argument to be made that this part should be tested in gitto and not here, but we are setting the webhook URL for these repositories from capstone, so that part should be tested here.
- Running Gitto in `network_mode="host"` could work. We would have to make some changes in gitto to let it listen on some port that is set in configuration. This doesn't seem like a big deal, and we could do it. One problem is that it again won't work with my setup where I am port-forwarding between a remote nomad deployment and my localhost.
- Running Gitto as a pytest fixture could work. I am hesitant to do this because we would have two deployments of Gitto -- one on Nomad and one on host. We would have to monkeypatch the configuration for a new Gitto URL. It also makes more sense to run these tests on a setup similar to what we'll have on production, since that's why we are writing integration tests like this.

I think we should go for the second option. I can work directly from my dev node until a better solution is found.

**Comment by nikochiko** [comment]
Created PR in Gitto for configuring port: https://github.com/pipalacademy/gitto/pull/5

**Comment by nikochiko** [comment]
@anandology monkeypatching the running capstone app cannot be done because it runs in a different process. I want to `monkeypatch` the `get_authenticated_user` method so that I can make a POST request to start user project.
I want to use `app.test_client`, as with normal flask testing, to work around this. It feels hacky at first to do this along side an app running, but it makes the running app's responsibility clear: to receive webhook calls only. If we decide to change how that is done later, we will only need to change that.

**Comment by anandology** [comment]
How about this? 
start the app with CAPSTONE_TEST_MODE=true and make it always return get_authenticated_user as guest when in that mode.

**Comment by nikochiko** [comment]
It would restrict what we can do inside tests (for example simulating multiple users would be harder than necessary).
What's wrong with using test client?

**Comment by anandology** [comment]
I'm suggesting to start the server with tetsmode to avoid the worry of authentication. You'll still use the test client to test it.

We have already spent too much time on getting the right test setup. I think we should strive for a good enough test setup that can be done in a reasonable time rather than the best setup that can handle every possible case.

**Comment by nikochiko** [comment]
We can directly set the session if we have access to test client.

**Comment by nikochiko** [comment]
done. new tests should be created with a new issue.
